File: go/0216M-组合总和III.go
/*
 * @Author: SukiEva
 * @Date: 2021-11-12 16:28:03
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
package main

func combinationSum3(k int, n int) [][]int {
	ans := make([][]int, 0)
	vis := make([]int, 0)
	var dfs func(p, sum int)
	dfs = func(p, sum int) {
		// 剪枝 ： 总和大于 n
		if sum > n {
			return
		}
		if len(vis) == k {
			if sum == n {
				tmp := make([]int, k)
				copy(tmp, vis)
				ans = append(ans, tmp)
			}
			return
		}
		for i := p; i <= 9-(k-len(vis))+1; i++ {
			vis = append(vis, i)
			dfs(i+1, sum+i)
			vis = vis[:len(vis)-1]
		}
	}
	dfs(1, 0)
	return ans
}

File: cpp/0216M-组合总和III.cpp
/*
 * @Author: SukiEva
 * @Date: 2021-12-26 13:57:02
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
#include <bits/stdc++.h>

#include "include/ListNode.h"
#include "include/TreeNode.h"

using namespace std;

class Solution {
   public:
    vector<vector<int>> ans;
    vector<int> tmp;

    void dfs(int k, int n, int idx, int sum) {
        if (sum > n) return;
        if (tmp.size() == k) {
            if (sum != n) return;
            ans.push_back(tmp);
            return;
        }
        for (int i = idx; i <= 9 - (k - tmp.size()) + 1; i++) {
            tmp.push_back(i);
            dfs(k, n, i + 1, sum + i);
            tmp.pop_back();
        }
    }

    vector<vector<int>> combinationSum3(int k, int n) {
        dfs(k, n, 1, 0);
        return ans;
    }
};